compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -100,10 +100,27 @@ object Applications {
       Nil
     }
 
+    def validUnapplySeqType(getTp: Type): Boolean = {
```

Comment:
I am wondering if we should do something along the line of (inspired by implementation of `isGetMatch`):
```scala
  def isValidUnapplySeqType(tp: Type, errorPos: Position)(implicit ctx: Context) =
    isGetMatch && {
      val getTpe = extractorMemberType(tp, nme.get, errorPos)
      val applyTpe = extractorMemberType(getTpe, nme.apply, errorPos)
      applyTpe.exists && {
        def dropTpe = extractorMemberType(getTpe, nme.applyTpe, errorPos)
        def toSeqTpe = extractorMemberType(getTpe, nme.toSeq, errorPos)
        def lengthCompareTpe = extractorMemberType(getTpe, nme.lengthCompare, errorPos)
        val lengthTpe = extractorMemberType(getTpe, nme.length, errorPos)

        (lengthCompareTpe.isRef(defn.BooleanClass) || lengthTpe.isRef(defn.IntClass)) && {
          val expectedSeqTpe = defn.SeqType.appliedTo(applyTpe)
          dropTpe <:< expectedSeqTpe && toSeqTpe <:< expectedSeqTpe
        }
      }
    }
```

Reply:
The original implementation is a concise way to implement the same logic. In the above, `extractorMemberType` doesn't work for methods with parameters, and we will need to check both NoType and param types of the methods.